Orchestrators

CNMauriello opened the pull request:
refactor and dummy main

Dabaduck committed:
Refactor PyPiHandler to use asyncio for concurrent fetching of package metadata and remove deprecated PyPiCLIENT.
Refactor package_metadata_fetcher and updatet API for cached object

Dabaduck committed:
Refactor main function and package metadata fetcher to improve logging and handle missing files; add package skipping logic and enhance dependency tree builder with root finding and full tree printing.

kbublij2 commented on src/analyzer/package_metadata_fetcher.py:
```diff
@@ -64,44 +70,47 @@ def build_package_metadata(self,
             # Extract all unique packages (keys + all values)
-            all_packages = set(graph.keys())
-            for deps in graph.values():
+            all_packages = set(_graph.keys())  # posso eliminare
```
This is an early build, notes are still welcome

kbublij2 commented on src/analyzer/package_metadata_fetcher.py:
```diff
@@ -52,10 +56,12 @@ def build_package_metadata(self,
             A list of PyPiMetadata objects containing package name, license, and link.
             Returns an empty list if file parsing fails.
         """
+        global _graph  # <-- ADD THIS LINE
```
Lack of this line breaks existing functionality, keeping that in.

kbublij2 commented on src/analyzer/package_metadata_fetcher.py:
```diff
@@ -64,44 +70,47 @@ def build_package_metadata(self,
-        )
-        for pkg, success in down_results.items():
-            LOGGER.info("Download %s: %s", pkg, success)
+        # down_results = self.repo_downloader.download_repos(
+        #    repo_urls=package_urls,
+        #    output_path=DOWNLOAD_DIRECTORY,
+        #    branch=DEFAULT_DOWNLOAD_BRANCH,
+        # )
+        # for pkg, success in down_results.items():
+        #    LOGGER.info("Download %s: %s", pkg, success)
```
This code covers example usage (seemingly), it's better left in

kbublij2 commented on src/infrastructure/pypi_client.py:
```diff
@@ -26,90 +25,131 @@ def get_source_links(self, packages_names: List[str],
-            license_expression = info.get("license_expression")
-            if isinstance(license_expression, str) and not license_expression.strip():
-                license_expression = None
-
-            # ------------------------------
-            # Final license selection
-            # ------------------------------
-            license_final = license_from_classifiers or \
-                license_simple or \
-                license_expression or \
-                "Unknown"
-
-            results[package] = {
-                'license': license_final,
-                #   'license_expression': license_expression,
-                'link': source_link
-            }
-            LOGGER.info("Fetched %s.json", package)
+        loop = asyncio.get_running_loop()
+
+        # Use a ThreadPoolExecutor to run synchronous requests non-blocking
+        with ThreadPoolExecutor() as executor:
+            tasks = []
+            for pkg in packages_names:
+                tasks.append(
+                    loop.run_in_executor(
+                        executor, self._process_single_package, pkg, timeout)
+                )
+
+            # Wait for all tasks to complete
+            completed_results = await asyncio.gather(*tasks)
+
+            for pkg_name, data in completed_results:
+                results[pkg_name] = data
 
         return results
```
Tests are being concerned separately for early stages.

kamilok04 committed:
Apply suggested fixes

Add missing docstrings
A slight exception handling improvement
Fix obsolete docstrings

kbublij2 reviewed (approved):
Looks good! Uses most, if not all, of the currently present project functionality.
